File: 2015/day06_1.py
# pylint: disable=missing-module-docstring, missing-function-docstring


import re
from typing import Any, Literal

type Instruction = tuple[
    Literal["on"] | Literal["off"] | Literal["toggle"], int, int, int, int
]
type Lights = list[list[bool]]


def solution(content: list[str]) -> Any:
    lights = [[False for _ in range(1000)] for _ in range(1000)]
    instructions = [parse(line) for line in content]
    for instruction in instructions:
        lights = act(instruction, lights)
    return len([True for row in lights for light in row if light])


def act(instruction: Instruction, lights: Lights) -> Lights:
    action, sx, sy, ex, ey = instruction
    copy = lights.copy()
    for x in range(sx, ex + 1):
        for y in range(sy, ey + 1):
            if action == "off":
                copy[x][y] = False
            elif action == "on":
                copy[x][y] = True
            else:
                copy[x][y] = not copy[x][y]
    return copy


def parse(line: str) -> Instruction:
    match = re.match(r"(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)", line)
    if match is None:
        raise ValueError(f"Invalid line {line}")

    return (
        (
            "on"
            if match[1] == "turn on"
            else "off" if match[1] == "turn off" else "toggle"
        ),
        int(match[2]),
        int(match[3]),
        int(match[4]),
        int(match[5]),
    )


def main() -> None:
    with open("day06.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()

File: 2015/day06-1.ts
type Lights = boolean[][];
type Instruction = {
  action: "on" | "off" | "toggle";
  sx: number;
  sy: number;
  ex: number;
  ey: number;
};

function solution(content: string) {
  let lights: Lights = Array.from({ length: 1000 }).map((_) =>
    Array.from({ length: 1000 }).map((_) => false)
  );
  content
    .split("\n")
    .map(parse)
    .forEach((instruction) => {
      lights = act(instruction, lights);
    });
  return lights
    .reduce((acc, cur) => [...acc, ...cur], [])
    .filter((item) => item).length;
}

function act(instruction: Instruction, lights: Lights): Lights {
  const { action, sx, sy, ex, ey } = instruction;
  const copy = lights.slice();
  for (let x = sx; x <= ex; x++) {
    for (let y = sy; y <= ey; y++) {
      if (action === "off") copy[x][y] = false;
      if (action === "on") copy[x][y] = true;
      if (action === "toggle") copy[x][y] = !copy[x][y];
    }
  }
  return copy;
}

function parse(line: string): Instruction {
  const match = line.match(
    /(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)/
  );
  if (!match) throw new Error(`Invalid line: ${line}`);
  return {
    action:
      match[1] === "turn on"
        ? "on"
        : match[1] === "turn off"
        ? "off"
        : "toggle",
    sx: Number(match[2]),
    sy: Number(match[3]),
    ex: Number(match[4]),
    ey: Number(match[5]),
  };
}

Deno.readTextFile("day06.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);
